### PR TITLE
Jv investigate getting processindicatorid without db lookup

### PR DESCRIPTION
**Pull request by JoukoVirtanen**
## Description

Instead of looking up processindicatorid in the process_indicator table using the process information in the plop object, determine the processindicatorid using the process information in the plop object using the same function that is used to determine the id for the process indicators. This is both more efficient and means that we get process information for listening endpoints when the listening endpoint arrives before the process indicator, without any complex mechanism for handling that case.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Ran e2e tests 200 times
- [x] Load testing
- [x] Performance testing

## Testing Performed

To test this properly the changes here were merged with other fixes. A PR for testing purposes was created here https://github.com/stackrox/stackrox/pull/5393

Performance tests are here https://docs.google.com/document/d/1Xv8eKZWiXLURGn9LVrLdo9o4-WJR49A0f7thInlBAFY/edit#heading=h.aptf438z8hjb

Load tests are here https://docs.google.com/document/d/1THsA2PohUYO_SYTT4uL2VcLmBKvVikW35x70nK9WP0E/edit#heading=h.wmfol0u90wc9

### CODE REVIEW

File: pkg/postgres/schema/process_indicators.go [hunk 1]
```diff
@@ -58,13 +58,13 @@ const (
 type ProcessIndicators struct {
```
Comment (dashrews78) [inline pkg/postgres/schema/process_indicators.go:58]: Just to clarify something to make sure we do not need a migration.  If memory serves these updates were NOT part of 3.74.  Is that correct?  If that is the case we do not need a migration to remove those indexes.

Comment (JoukoVirtanen) [reply pkg/postgres/schema/process_indicators.go:58]: Nothing related to processes listening on ports went into central for 3.74, except the feature flag. These updates were not a part of 3.74.

**Review by erthalion — COMMENTED** [review]
Few comments above.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 2]
```diff
@@ -116,10 +120,10 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
 		if !prevExists {
 			if val.CloseTimestamp != nil {
 				// We try to close a not existing Endpoint, something is wrong
-				log.Warnf("Found no matching PLOP to close for %s", key)
+				log.Warnf("Found no matching PLOP to close for %+v", val)
 			}
 
-			plopObjects = addNewPLOP(plopObjects, indicatorID, processInfo, val)
+			plopObjects = addNewPLOP(plopObjects, indicatorID, nil, val)
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go]: I brought in your commit from the other branch to fix this. There were a couple of minor mistakes in it that I fixed. Fix is confirmed by the unit tests.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 3]
```diff
@@ -63,40 +85,44 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
-		if indicator, ok := indicatorsMap[key]; ok {
-			indicatorID = indicator.GetId()
-			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
-		} else {
-			countMetrics.IncrementOrphanedPLOPCounter(val.GetClusterId())
-			log.Warnf("Found no matching indicators for %s", key)
-			processInfo = val.Process
-		}
-
+		indicatorID := getIndicatorIDForPlop(val)
```
Comment (dashrews78) [inline central/processlisteningonport/datastore/datastore_impl.go:113]: Do you need to handle the case where this call returns an empty string?  Or does processing work as desired in that case?

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 4]
```diff
@@ -130,24 +156,21 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
```
Comment (dashrews78) [inline central/processlisteningonport/datastore/datastore_impl.go:161]: same comment above as this returning empty and do we need to handle that case or not.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 5]
```diff
@@ -63,40 +85,44 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
-	// TODO ROX-14376: The next two calls, fetchIndicators and fetchExistingPLOPs, have to
-	// be done in a single join query fetching both ProcessIndicator and needed
-	// bits from PLOP.
-	indicatorsMap, indicatorIds, err := ds.fetchIndicators(ctx, normalizedPLOPs...)
-	if err != nil {
-		return err
+	allPLOPs := append(normalizedPLOPs, completedInBatch...)
+	indicatorIds := getIndicatorIdsForPlops(allPLOPs)
+
+	indicators, nonempty, err := ds.indicatorDataStore.GetProcessIndicators(ctx, indicatorIds)
```
[the remark this answers is not in the document]
Comment (erthalion) [reply central/processlisteningonport/datastore/datastore_impl.go:91]: `err == nil` is used a few lines below to actually prepare `indicatorsMap`. Is that what you asking about?

[the remark this answers is not in the document]
Comment (dashrews78) [reply central/processlisteningonport/datastore/datastore_impl.go:91]: yeah but we aren't doing anything if that call returns an `err != nil`.  Maybe we want to treat the case of `ds.indicatorDataStore.GetProcessIndicators` return an error the same as if it returns nothing or maybe we want to do something different when `err != nil`.  I was asking to make sure not handling `err != nil` was intentional.

[the remark this answers is not in the document]
Comment (erthalion) [reply central/processlisteningonport/datastore/datastore_impl.go:91]: Right, sorry for not communicating it more clearly. The idea was to treat it in the same way as empty response, not preparing `indicatorsMap` in such case. As this logic is purely for reducing amount of data we duplicate, every error scenario could be considered as not fatal, switch us to less performant path.

[the remark this answers is not in the document]
Comment (dashrews78) [reply central/processlisteningonport/datastore/datastore_impl.go:91]: got it.  That clarification makes sense to me.

[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:91]: So no changes are needed here?

[the remark this answers is not in the document]
Comment (dashrews78) [reply central/processlisteningonport/datastore/datastore_impl.go:91]: @JoukoVirtanen correct.  fine as is.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 6]
```diff
@@ -40,6 +40,28 @@ func newDatastoreImpl(
 	}
 }
 
+func getIndicatorIDForPlop(plop *storage.ProcessListeningOnPortFromSensor) string {
+	if plop == nil {
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:44]: Now it is there.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 7]
```diff
@@ -7,14 +7,14 @@ import (
 	"time"
 
 	"github.com/stackrox/rox/central/metrics"
-	countMetrics "github.com/stackrox/rox/central/metrics"
```
Comment (JoukoVirtanen) [inline central/processlisteningonport/datastore/datastore_impl.go:10]: Add metric back in

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 8]
```diff
@@ -486,7 +517,7 @@ func addNewPLOP(plopObjects []*storage.ProcessListeningOnPortStorage,
 	processInfo *storage.ProcessIndicatorUniqueKey,
 	value *storage.ProcessListeningOnPortFromSensor) []*storage.ProcessListeningOnPortStorage {
 
-	if value == nil {
+	if value == nil || indicatorID == "" {
```
Comment (JoukoVirtanen) [inline central/processlisteningonport/datastore/datastore_impl.go:520]: log.Warn here

File: central/processlisteningonport/datastore/datastore_impl.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:44]: Improve warning to say what the consequence of not being able to set id is.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 9]
```diff
@@ -63,40 +87,47 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
+	// Used to do best efforts of identifying orphaned PLOP. Note, that as
+	// usual, this could be done together with fetchExistingPLOPs in the same
+	// custom query.
+	if nonempty && err == nil {
+		for _, ind := range indicators {
+			indicatorsMap[ind.GetId()] = true
+		}
 	}
 
+	// TODO ROX-14376: This ticket might need to be changed, but is not completely obsolete
```
Comment (JoukoVirtanen) [inline central/processlisteningonport/datastore/datastore_impl.go:105]: Check ticket and see if it is still needed.

File: central/processlisteningonport/datastore/datastore_impl_test.go [hunk 10]
```diff
@@ -134,6 +133,20 @@ func (suite *PLOPDataStoreTestSuite) TestPLOPAdd() {
 		},
 	}
 
+	for _, indicator := range indicators {
```
Comment (JoukoVirtanen) [inline central/processlisteningonport/datastore/datastore_impl_test.go:136]: Multiple events with no matching process indicators. Two consecutive open and two consecutive close. Enable unit tests in CI.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 11]
```diff
@@ -63,40 +87,47 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
 		plopKey := getPlopKeyFromParts(val.GetProtocol(), val.GetPort(), indicatorID)
 
 		existingPLOP, prevExists := existingPLOPMap[plopKey]
 
+		// Best efforst to not duplicate data. If no process indicator with
+		// such an id exists, we deal with a potentially orphaned PLOP, and
+		// need to store the process information. Otherwise processInfo is nil
+		// and will not be stored.
+		if _, indicatorExists := indicatorsMap[indicatorID]; !indicatorExists {
+			processInfo = val.GetProcess()
```
Comment (JoukoVirtanen) [inline central/processlisteningonport/datastore/datastore_impl.go:128]: Add metric here.

Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:131]: Done

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 12]
```diff
@@ -130,24 +161,24 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:184]: Done

File: central/processlisteningonport/datastore/datastore_impl.go (hunk 7 above)
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go]: Done

File: central/processlisteningonport/datastore/datastore_impl.go (hunk 9 above)
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:106]: I don't think it is needed. I deleted the comment and closed the ticket.

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 13]
```diff
@@ -63,40 +87,47 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
-		if indicator, ok := indicatorsMap[key]; ok {
-			indicatorID = indicator.GetId()
-			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
-		} else {
-			countMetrics.IncrementOrphanedPLOPCounter(val.GetClusterId())
-			log.Warnf("Found no matching indicators for %s", key)
-			processInfo = val.Process
+		indicatorID := getIndicatorIDForPlop(val)
+		if indicatorID == "" {
+			continue
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:118]: Done

File: central/processlisteningonport/datastore/datastore_impl.go [hunk 14]
```diff
@@ -130,24 +161,24 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
...
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl.go:171]: Done

File: central/processlisteningonport/datastore/datastore_impl_test.go (hunk 10 above)
Comment (JoukoVirtanen) [reply central/processlisteningonport/datastore/datastore_impl_test.go:136]: Done

**Review by erthalion — APPROVED** [review]
Please update the PR description and the squashed commit message to describe the rationale behind this change.

**Review by dashrews78 — APPROVED** [review]

**Pull request merged**